File: services/api/src/creditops/domain/security_interests.py
"""Per-asset security interests and their perfection-item ledger (stage 9).

Master design section 5 giai đoạn 9.  Pure, frozen value objects plus the two
deterministic derivations the confirmation gate needs; NO I/O, NO valuation, NO
priority ranking.

SPEC CONTRACT faithfully encoded here:

- ONE ``SecurityInterest`` PER ASSET -- never a single case-wide boolean.  Each
  interest carries MANY ``SecurityPerfectionItem`` requirements, tracked
  individually (evidence, filing reference, effective/expiry date, status).
- ``SecurityAssetKind`` and ``PerfectionStatus`` are PROPOSED synthetic closed
  taxonomies; they MUST be reconfigured once an official source is supplied.
- NO LLM valuation: ``valuation_reference`` is a nullable POINTER to an external
  source/adapter, never a computed value.  There is NO ``priority_rank`` field
  anywhere -- the system never declares a final priority ranking; only the
  free-text ``notes_vi`` exists.
- A ``SecurityPerfectionItem`` advances ONLY through ``ALLOWED_ITEM_TRANSITIONS``
  (``PENDING`` -> ``EVIDENCE_ATTACHED`` -> ``COMPLETED``; ``PENDING`` ->
  ``NOT_REQUIRED_BY_HUMAN``; ``COMPLETED`` -> ``EXPIRED``).  A ``COMPLETED`` item
  MUST carry at least one evidence reference and record who/when completed it.

CONFIRMABLE DERIVATION (no vacuous paths):
``derive_perfection_confirmable`` is True IFF the case has at least one interest
AND every interest has at least one item AND every item is in a terminal-
satisfied state (``COMPLETED`` or ``NOT_REQUIRED_BY_HUMAN``).  An interest with
ZERO items is NOT confirmable; a case with ZERO interests is NOT confirmable.  An
explicit no-collateral path (a case that legitimately needs no security) is
PROPOSED / out of scope here and deliberately not modelled.

All data is synthetic and created solely for demonstration.
"""

from __future__ import annotations

from collections.abc import Sequence
from datetime import date, datetime
from enum import StrEnum
from typing import Self
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator

from creditops.domain.ids import CaseId

type SecurityInterestId = UUID
type SecurityPerfectionItemId = UUID


class SecurityAssetKind(StrEnum):
    """Closed PROPOSED synthetic set of secured-asset kinds (no official SHB
    taxonomy); reconfigure when an official source is supplied."""

    REAL_ESTATE = "REAL_ESTATE"
    VEHICLE = "VEHICLE"
    DEPOSIT = "DEPOSIT"
    RECEIVABLE = "RECEIVABLE"
    OTHER = "OTHER"


class PerfectionStatus(StrEnum):
    """Closed PROPOSED synthetic per-requirement status set.

    Advanced ONLY through ``ALLOWED_ITEM_TRANSITIONS``.  ``COMPLETED`` and
    ``NOT_REQUIRED_BY_HUMAN`` are the two terminal-satisfied states that a
    perfection requirement may rest in for the confirmation gate;
    ``EXPIRED`` is terminal but NOT satisfied (a lapsed registration reopens the
    requirement to human attention).
    """

    PENDING = "PENDING"
    EVIDENCE_ATTACHED = "EVIDENCE_ATTACHED"
    COMPLETED = "COMPLETED"
    NOT_REQUIRED_BY_HUMAN = "NOT_REQUIRED_BY_HUMAN"
    EXPIRED = "EXPIRED"


#: The ONLY permitted status transitions (mirrors the SQL trigger in
#: 202607180019_security_perfection.sql).  Every mapped value is the exact set of
#: statuses a requirement may move to from the key status; a terminal status maps
#: to the empty set.
ALLOWED_ITEM_TRANSITIONS: dict[PerfectionStatus, frozenset[PerfectionStatus]] = {
    PerfectionStatus.PENDING: frozenset(
        {PerfectionStatus.EVIDENCE_ATTACHED, PerfectionStatus.NOT_REQUIRED_BY_HUMAN}
    ),
    PerfectionStatus.EVIDENCE_ATTACHED: frozenset({PerfectionStatus.COMPLETED}),
    PerfectionStatus.COMPLETED: frozenset({PerfectionStatus.EXPIRED}),
    PerfectionStatus.NOT_REQUIRED_BY_HUMAN: frozenset(),
    PerfectionStatus.EXPIRED: frozenset(),
}

#: The terminal states in which a perfection requirement counts as SATISFIED for
#: the confirmation gate.  ``EXPIRED`` is deliberately excluded.
TERMINAL_SATISFIED_STATUSES: frozenset[PerfectionStatus] = frozenset(
    {PerfectionStatus.COMPLETED, PerfectionStatus.NOT_REQUIRED_BY_HUMAN}
)


def is_allowed_item_transition(
    current: PerfectionStatus, target: PerfectionStatus
) -> bool:
    """Whether ``current`` -> ``target`` is a permitted perfection transition."""

    return target in ALLOWED_ITEM_TRANSITIONS.get(current, frozenset())


class SecurityInterest(BaseModel):
    """One append-only security interest taken over ONE asset.

    ``valuation_reference`` is a pointer only; there is no monetary value and no
    priority ranking on this model by design.
    """

    model_config = ConfigDict(frozen=True, extra="forbid")

    id: SecurityInterestId
    case_id: CaseId
    case_version: int = Field(ge=1)
    asset_description_vi: str = Field(min_length=1, max_length=2000)
    asset_kind: SecurityAssetKind
    owner_name_vi: str | None = Field(default=None, min_length=1, max_length=500)
    #: POINTER ONLY to an external valuation source/adapter (nullable). NEVER a
    #: computed or LLM-produced value.
    valuation_reference: str | None = Field(default=None, min_length=1, max_length=500)
    notes_vi: str | None = Field(default=None, min_length=1, max_length=4000)
    created_by: UUID


class SecurityPerfectionItem(BaseModel):
    """One perfection requirement of one interest, tracked individually.

    A ``COMPLETED`` requirement MUST carry at least one evidence reference and
    record ``completed_by`` / ``completed_at`` (mirrors the SQL CHECK).  Other
    statuses are unconstrained on those fields.
    """

    model_config = ConfigDict(frozen=True, extra="forbid")

    id: SecurityPerfectionItemId
    interest_id: SecurityInterestId
    requirement_vi: str = Field(min_length=1, max_length=2000)
    status: PerfectionStatus = PerfectionStatus.PENDING
    evidence_refs: tuple[str, ...] = ()
    filing_reference: str | None = Field(default=None, min_length=1, max_length=500)
    effective_date: date | None = None
    expiry_date: date | None = None
    completed_by: UUID | None = None
    completed_at: datetime | None = None

    @model_validator(mode="after")
    def _completed_requires_evidence(self) -> Self:
        if self.status is PerfectionStatus.COMPLETED:
            if not self.evidence_refs:
                raise ValueError(
                    "a COMPLETED perfection item requires at least one evidence "
                    "reference"
                )
            if any(not ref.strip() for ref in self.evidence_refs):
                raise ValueError("every evidence reference must be non-empty")
            if self.completed_by is None or self.completed_at is None:
                raise ValueError(
                    "a COMPLETED perfection item must record completed_by and "
                    "completed_at"
                )
        return self


class SecurityInterestWithItems(BaseModel):
    """An interest paired with its perfection items (read model for derivation)."""

    model_config = ConfigDict(frozen=True, extra="forbid")

    interest: SecurityInterest
    items: tuple[SecurityPerfectionItem, ...] = ()


class PerfectionBlockers(BaseModel):
    """The structured reason a case is (not yet) perfection-confirmable.

    ``blocking_item_ids`` are perfection items NOT in a terminal-satisfied state;
    ``interests_without_items`` are interests carrying zero requirements (each an
    independent blocker).  ``has_interests`` is False for a case with no
    interests at all.  ``confirmable`` is True only when every blocking channel is
    clear.
    """

    model_config = ConfigDict(frozen=True, extra="forbid")

    has_interests: bool
    interests_without_items: tuple[UUID, ...] = ()
    blocking_item_ids: tuple[UUID, ...] = ()

    @property
    def confirmable(self) -> bool:
        return (
            self.has_interests
            and not self.interests_without_items
            and not self.blocking_item_ids
        )


def derive_perfection_blockers(
    interests_with_items: Sequence[SecurityInterestWithItems],
) -> PerfectionBlockers:
    """Pure structural derivation of everything blocking confirmation.

    No vacuous paths: a case with zero interests, or any interest with zero
    items, is a blocker; so is any item not in ``TERMINAL_SATISFIED_STATUSES``.
    """

    interests_without_items: list[UUID] = []
    blocking_item_ids: list[UUID] = []
    for entry in interests_with_items:
        if not entry.items:
            interests_without_items.append(entry.interest.id)
            continue
        for item in entry.items:
            if item.status not in TERMINAL_SATISFIED_STATUSES:
                blocking_item_ids.append(item.id)
    return PerfectionBlockers(
        has_interests=bool(interests_with_items),
        interests_without_items=tuple(interests_without_items),
        blocking_item_ids=tuple(blocking_item_ids),
    )


def derive_perfection_confirmable(
    interests_with_items: Sequence[SecurityInterestWithItems],
) -> bool:
    """Whether the security-perfection gate MAY be confirmed right now.

    True IFF the case has >=1 interest AND every interest has >=1 item AND every
    item is ``COMPLETED`` or ``NOT_REQUIRED_BY_HUMAN``.  Never vacuously true.
    """

    return derive_perfection_blockers(interests_with_items).confirmable


__all__ = [
    "ALLOWED_ITEM_TRANSITIONS",
    "TERMINAL_SATISFIED_STATUSES",
    "PerfectionBlockers",
    "PerfectionStatus",
    "SecurityAssetKind",
    "SecurityInterest",
    "SecurityInterestId",
    "SecurityInterestWithItems",
    "SecurityPerfectionItem",
    "SecurityPerfectionItemId",
    "derive_perfection_blockers",
    "derive_perfection_confirmable",
    "is_allowed_item_transition",
]

File: apps/web/lib/api/security-interests.ts
// Self-contained API bindings for the stage-9 Security Perfection ledger screen
// ("Hoàn thiện biện pháp bảo đảm"). Mirrors lib/api/client.ts conventions but
// shares no mutable state with it.
//
// Backend truth mirrored here: services/api/src/creditops/api/security_interests.py
// and services/.../domain/security_interests.py.
//   GET  /api/v1/cases/{caseId}/security-interests                              -> Ledger
//   POST /api/v1/cases/{caseId}/security-interests                              -> 201 Interest
//   POST /api/v1/cases/{caseId}/security-interests/{interestId}/items           -> 201 Item (starts PENDING)
//   POST /api/v1/cases/{caseId}/security-interests/items/{itemId}/transition    -> Item (closed status graph)
//   POST /api/v1/cases/{caseId}/security-interests/confirm                      -> Confirmation (HG_SECURITY_PERFECTION_CONFIRMED)
//
// ONE interest per asset, each carrying MANY perfection requirements tracked
// individually. There is NO valuation and NO priority ranking. Confirmation is
// fail-closed: every interest must carry >=1 requirement and every requirement
// must be terminally satisfied.

import { ApiClientError } from "./client";

const BFF_BASE_URL = "/api/creditops";
const CSRF_COOKIE_NAME = "__Host-creditops-csrf";
const CSRF_HEADER_NAME = "x-creditops-csrf";

// --- Domain enums (mirror services/.../domain/security_interests.py) ---

export type GateStatus = "OPEN" | "SATISFIED";

export type SecurityAssetKind =
  | "REAL_ESTATE"
  | "VEHICLE"
  | "DEPOSIT"
  | "RECEIVABLE"
  | "OTHER";

export type PerfectionStatus =
  | "PENDING"
  | "EVIDENCE_ATTACHED"
  | "COMPLETED"
  | "NOT_REQUIRED_BY_HUMAN"
  | "EXPIRED";

export const SECURITY_ASSET_KINDS: readonly SecurityAssetKind[] = [
  "REAL_ESTATE",
  "VEHICLE",
  "DEPOSIT",
  "RECEIVABLE",
  "OTHER",
];

// The ONLY permitted per-requirement transitions (mirrors ALLOWED_ITEM_TRANSITIONS
// and the SQL trigger). A terminal status maps to an empty set; the UI derives
// the choices from the current status so it can never OFFER a forbidden move.
export const ALLOWED_ITEM_TRANSITIONS: Record<PerfectionStatus, readonly PerfectionStatus[]> = {
  PENDING: ["EVIDENCE_ATTACHED", "NOT_REQUIRED_BY_HUMAN"],
  EVIDENCE_ATTACHED: ["COMPLETED"],
  COMPLETED: ["EXPIRED"],
  NOT_REQUIRED_BY_HUMAN: [],
  EXPIRED: [],
};

// The two terminal states in which a requirement counts as satisfied for the
// confirmation gate (EXPIRED is deliberately excluded).
export const TERMINAL_SATISFIED_STATUSES: readonly PerfectionStatus[] = [
  "COMPLETED",
  "NOT_REQUIRED_BY_HUMAN",
];

// A COMPLETED requirement MUST carry at least one evidence reference (mirrors the
// domain model validator + SQL CHECK); the transition form enforces this before
// it ever calls the API.
export const COMPLETED_REQUIRES_EVIDENCE = true;

// --- Response shapes ---

export interface PerfectionItem {
  id: string;
  interestId: string;
  requirement: string;
  status: PerfectionStatus | string;
  evidenceRefs: string[];
  filingReference: string | null;
  effectiveDate: string | null;
  expiryDate: string | null;
  completedBy: string | null;
  completedAt: string | null;
  createdAt: string;
}

export interface SecurityInterest {
  id: string;
  caseId: string;
  caseVersion: number;
  assetDescription: string;
  assetKind: SecurityAssetKind | string;
  ownerName: string | null;
  valuationReference: string | null;
  notes: string | null;
  createdBy: string;
  createdAt: string;
}

export interface InterestWithItems {
  interest: SecurityInterest;
  items: PerfectionItem[];
}

export interface SecurityLedger {
  interests: InterestWithItems[];
}

export interface SecurityConfirmation {
  gateType: string;
  status: GateStatus | string;
  dispositionRef: string;
}

export interface CreateInterestInput {
  assetDescription: string;
  assetKind: SecurityAssetKind;
  ownerName?: string;
  valuationReference?: string;
  notes?: string;
}

export interface AddItemInput {
  requirement: string;
  evidenceRefs?: string[];
  filingReference?: string;
  effectiveDate?: string;
  expiryDate?: string;
}

export interface TransitionItemInput {
  toStatus: PerfectionStatus;
  rationale?: string;
  evidenceRefs?: string[];
  filingReference?: string;
  effectiveDate?: string;
  expiryDate?: string;
}

// --- Defensive parsing ---

function asRecord(value: unknown): Record<string, unknown> {
  return typeof value === "object" && value !== null ? (value as Record<string, unknown>) : {};
}

function str(value: unknown): string {
  return typeof value === "string" ? value : value == null ? "" : String(value);
}

function strOrNull(value: unknown): string | null {
  return typeof value === "string" ? value : null;
}

function num(value: unknown): number {
  return typeof value === "number" && Number.isFinite(value) ? value : 0;
}

function strArray(value: unknown): string[] {
  return Array.isArray(value) ? value.map(str) : [];
}

function parseItem(value: unknown): PerfectionItem {
  const raw = asRecord(value);
  return {
    id: str(raw.id),
    interestId: str(raw.interestId),
    requirement: str(raw.requirement),
    status: str(raw.status),
    evidenceRefs: strArray(raw.evidenceRefs),
    filingReference: strOrNull(raw.filingReference),
    effectiveDate: strOrNull(raw.effectiveDate),
    expiryDate: strOrNull(raw.expiryDate),
    completedBy: strOrNull(raw.completedBy),
    completedAt: strOrNull(raw.completedAt),
    createdAt: str(raw.createdAt),
  };
}

function parseInterest(value: unknown): SecurityInterest {
  const raw = asRecord(value);
  return {
    id: str(raw.id),
    caseId: str(raw.caseId),
    caseVersion: num(raw.caseVersion),
    assetDescription: str(raw.assetDescription),
    assetKind: str(raw.assetKind),
    ownerName: strOrNull(raw.ownerName),
    valuationReference: strOrNull(raw.valuationReference),
    notes: strOrNull(raw.notes),
    createdBy: str(raw.createdBy),
    createdAt: str(raw.createdAt),
  };
}

function parseInterestWithItems(value: unknown): InterestWithItems {
  const raw = asRecord(value);
  return {
    interest: parseInterest(raw.interest),
    items: Array.isArray(raw.items) ? raw.items.map(parseItem) : [],
  };
}

export function parseSecurityLedger(value: unknown): SecurityLedger {
  const raw = asRecord(value);
  return {
    interests: Array.isArray(raw.interests)
      ? raw.interests.map(parseInterestWithItems)
      : [],
  };
}

export function parseSecurityInterest(value: unknown): SecurityInterest {
  return parseInterest(value);
}

export function parsePerfectionItem(value: unknown): PerfectionItem {
  return parseItem(value);
}

export function parseSecurityConfirmation(value: unknown): SecurityConfirmation {
  const raw = asRecord(value);
  return {
    gateType: str(raw.gateType),
    status: str(raw.status),
    dispositionRef: str(raw.dispositionRef),
  };
}

// --- Pure derivation (mirrors derive_perfection_blockers) ---

export interface PerfectionBlockers {
  hasInterests: boolean;
  interestsWithoutItems: string[];
  blockingItemIds: string[];
  confirmable: boolean;
}

// Structural client-side hint (the server is authoritative on confirm). No
// vacuous paths: zero interests, an interest with zero items, or any item not
// terminally satisfied is a blocker.
export function derivePerfectionBlockers(
  interests: readonly InterestWithItems[],
): PerfectionBlockers {
  const interestsWithoutItems: string[] = [];
  const blockingItemIds: string[] = [];
  for (const entry of interests) {
    if (entry.items.length === 0) {
      interestsWithoutItems.push(entry.interest.id);
      continue;
    }
    for (const item of entry.items) {
      if (!(TERMINAL_SATISFIED_STATUSES as readonly string[]).includes(item.status)) {
        blockingItemIds.push(item.id);
      }
    }
  }
  const hasInterests = interests.length > 0;
  return {
    hasInterests,
    interestsWithoutItems,
    blockingItemIds,
    confirmable:
      hasInterests && interestsWithoutItems.length === 0 && blockingItemIds.length === 0,
  };
}

export function allowedItemTransitions(status: string): readonly PerfectionStatus[] {
  return ALLOWED_ITEM_TRANSITIONS[status as PerfectionStatus] ?? [];
}

// --- Client ---

type Fetcher = typeof fetch;
type CsrfTokenProvider = () => string | null;

function readBrowserCsrfToken(): string | null {
  if (typeof document === "undefined") return null;
  for (const part of document.cookie.split(";")) {
    const index = part.indexOf("=");
    if (index < 0 || part.slice(0, index).trim() !== CSRF_COOKIE_NAME) continue;
    try {
      return decodeURIComponent(part.slice(index + 1).trim());
    } catch {
      return null;
    }
  }
  return null;
}

async function parseJson(response: Response): Promise<unknown> {
  const contentType = response.headers.get("content-type") ?? "";
  if (!contentType.includes("application/json")) return null;
  try {
    return await response.json();
  } catch {
    return null;
  }
}

function parseApiError(
  body: unknown,
): { code: string; messageVi: string; retryable: boolean } | null {
  if (typeof body !== "object" || body === null) return null;
  const raw = body as Record<string, unknown>;
  if (typeof raw.code !== "string") return null;
  return {
    code: raw.code,
    messageVi: typeof raw.messageVi === "string" ? raw.messageVi : "",
    retryable: typeof raw.retryable === "boolean" ? raw.retryable : false,
  };
}

export class SecurityInterestsApiClient {
  private readonly baseUrl: string;

  constructor(
    baseUrl = BFF_BASE_URL,
    private readonly fetcher: Fetcher = (input, init) => fetch(input, init),
    private readonly csrfTokenProvider: CsrfTokenProvider = readBrowserCsrfToken,
  ) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
  }

  async getLedger(caseId: string): Promise<SecurityLedger> {
    return parseSecurityLedger(
      await this.request(
        `/api/v1/cases/${encodeURIComponent(caseId)}/security-interests`,
      ),
    );
  }

  async createInterest(
    caseId: string,
    input: CreateInterestInput,
  ): Promise<SecurityInterest> {
    return parseSecurityInterest(
      await this.request(
        `/api/v1/cases/${encodeURIComponent(caseId)}/security-interests`,
        { method: "POST", body: JSON.stringify(compact(input)) },
      ),
    );
  }

  async addItem(
    caseId: string,
    interestId: string,
    input: AddItemInput,
  ): Promise<PerfectionItem> {
    return parsePerfectionItem(
      await this.request(
        `/api/v1/cases/${encodeURIComponent(caseId)}/security-interests/${encodeURIComponent(
          interestId,
        )}/items`,
        { method: "POST", body: JSON.stringify(compact(input)) },
      ),
    );
  }

  async transitionItem(
    caseId: string,
    itemId: string,
    input: TransitionItemInput,
  ): Promise<PerfectionItem> {
    return parsePerfectionItem(
      await this.request(
        `/api/v1/cases/${encodeURIComponent(caseId)}/security-interests/items/${encodeURIComponent(
          itemId,
        )}/transition`,
        { method: "POST", body: JSON.stringify(compact(input)) },
      ),
    );
  }

  async confirm(caseId: string, rationale: string): Promise<SecurityConfirmation> {
    return parseSecurityConfirmation(
      await this.request(
        `/api/v1/cases/${encodeURIComponent(caseId)}/security-interests/confirm`,
        { method: "POST", body: JSON.stringify({ rationale }) },
      ),
    );
  }

  private async request(path: string, init: RequestInit = {}): Promise<unknown> {
    const headers = new Headers(init.headers);
    headers.set("Accept", "application/json");
    if (init.body !== undefined) headers.set("Content-Type", "application/json");
    if (isMutation(init.method)) {
      const csrfToken = this.csrfTokenProvider();
      if (csrfToken) headers.set(CSRF_HEADER_NAME, csrfToken);
    }

    const response = await this.fetcher(`${this.baseUrl}${path}`, {
      ...init,
      headers,
      credentials: "include",
      cache: "no-store",
    });
    const body = await parseJson(response);
    if (!response.ok) {
      const apiError = parseApiError(body);
      throw new ApiClientError(
        response.status,
        apiError?.code ?? "REQUEST_FAILED",
        apiError?.messageVi || "Yêu cầu không thành công.",
        apiError?.retryable ?? response.status >= 500,
      );
    }
    return body;
  }
}

// Drops undefined / empty-array optional fields so the BFF's exact-keys check
// never sees a key the officer did not fill in.
function compact(input: object): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(input)) {
    if (value === undefined) continue;
    if (Array.isArray(value) && value.length === 0) continue;
    out[key] = value;
  }
  return out;
}

function isMutation(method: string | undefined): boolean {
  return method !== undefined && !["GET", "HEAD"].includes(method.toUpperCase());
}

export const securityInterestsApi = new SecurityInterestsApiClient();

const GENERIC_MESSAGES = new Set(["Không thể hoàn tất yêu cầu.", "Yêu cầu không thành công."]);

export function getSecurityError(error: unknown): string {
  if (error instanceof ApiClientError) {
    if (error.message && !GENERIC_MESSAGES.has(error.message)) {
      return error.message;
    }
    switch (error.status) {
      case 401:
        return "Phiên làm việc đã hết hạn. Vui lòng đăng nhập lại.";
      case 403:
        return "Bạn không có thẩm quyền thao tác trên biện pháp bảo đảm.";
      case 404:
        return "Không tìm thấy hồ sơ, biện pháp bảo đảm hoặc yêu cầu. Vui lòng tải lại.";
      case 409:
        return "Không thể hoàn tất: trạng thái đã thay đổi hoặc điều kiện chưa đủ. Vui lòng tải lại.";
      case 422:
        return "Dữ liệu chưa hợp lệ. Vui lòng kiểm tra và thử lại.";
      case 503:
        return "Dịch vụ hoàn thiện bảo đảm chưa sẵn sàng. Vui lòng thử lại sau ít phút.";
    }
  }
  return "Không thể hoàn tất yêu cầu. Vui lòng thử lại.";
}

// --- Display labels ---

export const GATE_STATUS_LABELS: Record<GateStatus, string> = {
  OPEN: "Đang chờ",
  SATISFIED: "Đạt",
};

export const ASSET_KIND_LABELS: Record<SecurityAssetKind, string> = {
  REAL_ESTATE: "Bất động sản",
  VEHICLE: "Phương tiện",
  DEPOSIT: "Tiền gửi",
  RECEIVABLE: "Khoản phải thu",
  OTHER: "Khác",
};

export const PERFECTION_STATUS_LABELS: Record<PerfectionStatus, string> = {
  PENDING: "Chờ xử lý",
  EVIDENCE_ATTACHED: "Đã đính kèm bằng chứng",
  COMPLETED: "Đã hoàn thiện",
  NOT_REQUIRED_BY_HUMAN: "Không yêu cầu (do người phụ trách)",
  EXPIRED: "Đã hết hiệu lực",
};

// Precise per-target action labels (never a generic verb). Each states the exact
// effect of the transition on the requirement.
export const ITEM_TRANSITION_LABELS: Record<PerfectionStatus, string> = {
  PENDING: "Đưa về chờ xử lý",
  EVIDENCE_ATTACHED: "Ghi nhận đã đính kèm bằng chứng",
  COMPLETED: "Ghi nhận hoàn thiện bảo đảm",
  NOT_REQUIRED_BY_HUMAN: "Xác định không yêu cầu",
  EXPIRED: "Ghi nhận hết hiệu lực",
};

export const UNSUPPORTED_ENUM_LABEL = "Trạng thái chưa được hỗ trợ";

export function labelOrUnsupported<K extends string>(
  map: Record<K, string>,
  key: string,
): string {
  return (map as Record<string, string>)[key] ?? UNSUPPORTED_ENUM_LABEL;
}

export function shortId(value: string | null | undefined): string {
  if (!value) return "—";
  return value.length > 12 ? `${value.slice(0, 8)}…` : value;
}

export function formatDateTime(iso: string): string {
  const date = new Date(iso);
  if (Number.isNaN(date.getTime())) return iso || "—";
  return date.toLocaleString("vi-VN");
}

export function formatDate(iso: string | null): string {
  if (!iso) return "—";
  const date = new Date(iso);
  if (Number.isNaN(date.getTime())) return iso;
  return date.toLocaleDateString("vi-VN");
}
